### PR TITLE
Created the Code and Terraform for the Daily-Report

### PR DESCRIPTION
**Pull request by CharliedeanCS**
Code : 

I have created a script for the daily report. This script contains a lambda handler which returns a body of html. This body of html creates a report of data from the previous days worth of data in the database. 

The script can handle changes to the database and can scale to any size required no matter the amount of data fed into the database the previous day.

This script also send a html document to a S3 bucket of the report to ensure it is stored for the long-term.

All code in the daily report has been tested through the test_report.py file. These tests ensure each function works as intended. 

Terraform:

I have created the terraform code for 4 services:
     - `Lambda Function`
      - A Lambda function that sends a html report of the daily rides to a S3 bucket and returns the html body in the lambda 
         handler return.
     - `State Function`
      - A State Function that runs the above Lambda and uses the html body sent from the Lambda. The body is used for a 
         SES V2 send email that sends an email to the ceo of deloton about the daily rides.
     - `EventBridge Schedule`
          An EventBridge Schedule that occurs everyday at 9am and activates the State Function to send the previous days data inside a report to the deloton ceo.

Dockerfile:

A dockerfile has also been created that creates a image and container for the daily report allowing it to be uploaded to AWS ECR.

Both the Terraform and Daily report README have also been updated to explain the updated code.

### CODE REVIEW

File: daily_report/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+WORKDIR ${LAMBDA_TASK_ROOT}
+
+COPY requirements.txt .
+
+RUN pip install -r requirements.txt
+
+COPY report.py .
+
+CMD [ "report.handler" ]
```
Comment (lcawood) [inline daily_report/Dockerfile:11]: All looks good to me, we can see if this works as intended later of course

File: daily_report/report.py [hunk 2]
```diff
@@ -1,9 +1,272 @@
...
+
+    except Exception as e:
+        return {
+            'statusCode': 404,
+            'body': json.dumps(str(e))
+        }
 
 if __name__ == "__main__":
-    report()
+    print(handler())
```
Comment (lcawood) [inline daily_report/report.py:272]: Having seen the report itself, I'm happy with the content and format :)

File: daily_report/test_report.py [hunk 3]
```diff
@@ -1,8 +1,215 @@
...
+
+@patch('report.previous_day_from_database')
+@patch('report.get_s3_client')
+@patch('report.get_database_connection')
+@patch('report.load_dotenv')
+@patch('report.sql_select_all_useful_data')
+@patch('report.extract_report_data')
+def test_handler_returns_error_message(mock_extract_report_data,mock_sql_select_all_useful_data,
+                              mock_load_dotenv,mock_get_database_connection,
+                              mock_get_s3_client,mock_previous_day_from_database):
```
Comment (lcawood) [inline daily_report/test_report.py:205]: Good test!

File: terraform/main.tf [hunk 4]
```diff
@@ -47,4 +47,296 @@ resource "aws_db_instance" "c9_velo_deloton" {
...
+      AWS_SECRET_ACCESS = "${var.aws_secret_access_key}"
+    }
+}
+}
+
+
+# Report : Step Function Permissions
+
+resource "aws_iam_role" "iam_for_sfn" {
+  name = "stepFunctionSampleStepFunctionExecutionIAM"
```
Comment (lcawood) [inline terraform/main.tf:129]: Looks like you've terraformed everything we need for this commit including all roles etc, nice.

**Review by lcawood — APPROVED** [review]

File: daily_report/report.py [hunk 5]
```diff
@@ -1,9 +1,272 @@
...
+A script then when ran creates a html string a document that show
+a daily report of the rides from the previous day.
+The html body is returned in a lambda handler.
+The html document is sent to a S3 Bucket.
+"""
 
-def report():
-    """All code retaining to the report"""
-    return 0
+# pylint: disable=C0301
```
Comment (CharliedeanCS) [inline daily_report/report.py:8]: Test

**Pull request merged**